CIP-0072? | DApp Registration & Discovery (#355)
cardano-foundation/CIPs: CIP-0072/README.md

---
CIP: 72
Title: Cardano DApp Registration
Authors: Bruno Martins <bruno.martins@iohk.io>
Status: Draft
Type: Standards
Created: 2022-10-18
License: CC-BY-4.0
---

## **Abstract**
DApp developers do not have a standardised method to record immutable, persistent claims about their dApp(s) that their users can verify. A dApp developers needs to "register" their dApp by providing a set of claims about their dApp(s) that can be verified by the user. This CIP describes a standardised method for dApp developers to register their dApp(s) and for users to verify the claims made by the dApp developer.

This proposal aims to standardise the process of dApp registration and verification, and to provide a set of claims that dApp developers can use to register their dApp(s).

## **Motivation**

> DApps can express a plethora of information. Some of this information could be claims about who the developer is, what is the dApp's associated metadata are, and more. This data lacks standarisation, persistence, and immutabulity. Data without these features, means that dApp users cannot verify if the dApp's expressed information is consistent across time. The goal of this CIP is to formalise how dApps register their information with a new transaction metadata format that can record the dApp's metadata, ownership, and potentially developer's identity. This formalisation means dApp users can verify if the data expressed by a dApp is consistent with what was registered on-chain.

Reviewer: typo: immutability

Author: thank you :+1: typo fixed

Also having this formalisation facilitates any actor in the ecosystem to index and query this data, and provide a better user experience when it comes to dApp discovery and usage.

## **Specification**

### **On-chain dApp Registration certificate**

```json
{
	"subject": "d684512ccb313191dd08563fd8d737312f7f104a70d9c72018f6b0621ea738c5b8213c8365b980f2d8c48d5fbb2ec3ce642725a20351dbff9861ce9695ac5db8",
	"rootHash": "8c4e9eec512f5f277ab811ba75c991d51600c80003e892e601c6b6c19aaf8a33",

>     "cip26": ["http://somelocation.io/", "http://somelocation2.io/"],

Reviewer: @aleks-djuric

	"did": "did:prism:f2019bd31a8530fb67c6d81da758dfa9a65be09d835cf2cd361d595a8858301d",
	"signature": {
		"r": "5114674f1ce8a2615f2b15138944e5c58511804d72a96260ce8c587e7220daa90b9e65b450ff49563744d7633b43a78b8dc6ec3e3397b50080",
		"s": "a15f06ce8005ad817a1681a4e96ee6b4831679ef448d7c283b188ed64d399d6bac420fadf33964b2f2e0f2d1abd401e8eb09ab29e3ff280600",
		"algo": "Ed25519−EdDSA",
		"pub": "b7a3c12dc0c8c748ab07525b701122b88bd78f600c76342d27f25e5f92444cde"
	}
}
```

### Properties

> *`subject`*: Identifier of the claim subject (i.e dapp). A UTF-8 encoded string. This uniquess of this property cannot be guaranteed by the protocol and multiple claims for the same subject may exist therefore it is required to exist some mechanism to assert trust in the *veracity* of this property. This can be done by knowning verifying the signature against a trusted / known public key or by using Decentralized Identifiers (DIDs) as described in the [DID Specification](https://www.w3.org/TR/did-core/).

Reviewer: I discussed similar topic with George from mLabs. Creator of a dApp needs to create some key and then all updates to dApp metadata (separate CIP coming from mLabs on this) should be with a subkey. By checking subkey of a key we can actually verify if the owner is the same.

One of the things we are hoping to work on is CIP-26 compatible dApp registry. Generally it is not that hard to do (read only access) but tricky part is testing whether dApp update has happened via the creator of the project or somebody else. Now if it is somebody else they could be malicious and update information maliciously, e.g. even pointing to the scam website.

Author: Thank you @matiwinnetou. 

I've added an extra `type` (register, update), perhaps this would help. Over time per subject we can trace the actions of the owner with different registrations. We can add extra actions if necessary. Open to discuss alternatives

*`rootHash`*: The hash of the metadata tree object. This hash is used by clients to verify the integrity of the metadata tree object. When reading a metadata tree object, the client should calculate the hash of the object and compare it with the `rootHash` property. If the two hashes don't match, the client should discard the object. The metadata tree object is a JSON object that contains the dApp's metadata. The metadata tree object is described in the next section.

*`cip26`*: An array of URLs that point to the location of the dApp's metadata. The client should try to fetch the metadata from the first URL in the array. If the client fails to fetch the metadata from the first URL, it should try the second URL, and so on. If the client fails to fetch the metadata from all URLs, it should consider the metadata to be invalid.

The format of the metadata tree object is described in the [CIP-26](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026) specification.

> *`did`*(optional): The DID of the dApp's developer. The DID is used to verify the authenticity of the certificate. The client should use the DID to fetch the developer's public key from the Cardano blockchain. The client should then use the public key to verify the signature of the certificate. This field is *optional* and can be used by identity centric dApps to verify the authenticity of the developer.

Reviewer: 👍

Also having this formalisation facilitates any actor in the ecosystem to index and query this data, and provide a better user experience when it comes to dApp discovery and usage.

## **Specification**

### **On-chain dApp Registration certificate**

```json
{
	"subject": "d684512ccb313191dd08563fd8d737312f7f104a70d9c72018f6b0621ea738c5b8213c8365b980f2d8c48d5fbb2ec3ce642725a20351dbff9861ce9695ac5db8",
	"rootHash": "8c4e9eec512f5f277ab811ba75c991d51600c80003e892e601c6b6c19aaf8a33",
  "type": "REGISTER",
  "cip26": ["http://somelocation.io/", "http://somelocation2.io/"],
	"did": "did:ada:f2019bd31a8530fb67c6d81da758dfa9a65be09d835cf2cd361d595a8858301d",
	"signature": {
		"r": "5114674f1ce8a2615f2b15138944e5c58511804d72a96260ce8c587e7220daa90b9e65b450ff49563744d7633b43a78b8dc6ec3e3397b50080",
		"s": "a15f06ce8005ad817a1681a4e96ee6b4831679ef448d7c283b188ed64d399d6bac420fadf33964b2f2e0f2d1abd401e8eb09ab29e3ff280600",
		"algo": "Ed25519−EdDSA",
		"pub": "b7a3c12dc0c8c748ab07525b701122b88bd78f600c76342d27f25e5f92444cde"
	}
}

> ```

Reviewer 2: 🚧 I suggest to add a schema version field, something like `dapps_registration_version: 1.0`.  We can anticipate that there will be newer schema versions and Dapps Stores will need to support multiple formats over time as well as potentially deprecate or ignore older ones.

*`type`*: The type of the claim. A UTF-8 encoded string. The value of this property can have the value `REGISTER` or `UPDATE`. The `REGISTER` type is used to register a new dApp. The `UPDATE` type is used to update the metadata of an existing dApp.

*`rootHash`*: The hash of the metadata entire tree object. This hash is used by clients to verify the integrity of the metadata tree object. When reading a metadata tree object, the client should calculate the hash of the object and compare it with the `rootHash` property. If the two hashes don't match, the client should discard the object. The metadata tree object is a JSON object that contains the dApp's metadata. The metadata tree object is described in the next section.

This hash is calculated by taking the entire metadata tree object, ordering the keys in the object alphanumerically, and then hashing the resulting JSON string using the blake2b hashing algorithm. The hash is encoded as a hex string.

> *`cip26`*: An array of URLs that point to the location of the dApp's metadata. The client should try to fetch the metadata from the first URL in the array. If the client fails to fetch the metadata from the first URL, it should try the second URL, and so on. If the client fails to fetch the metadata from all URLs, it should consider the metadata to be invalid.

Reviewer 2: ❓ Would it be preferable to have an array of `{subject}` instead of URLs?  This would keep the spirit of CIP-26 `We therefore believe that it is critical that we allow users to choose which metadata server (or servers) they refer to.`.  In this context, clients can be dapps registrants or dapps certifiers.  Dapps certifiers could publish the list of metadata servers they trust and dapps developers seeking for certification would have to target the list of metadata servers used by the desired dapps certifiers.

Moreover, in the event that a metadata server operator need to update its base URL, developers could have to re-issue registrations and certifiers would have to certify again.

Author: Thanks @flyerman . I think that's a great idea. In that way we won't be "marrying" the registration / certifications to a particular spec (I.e CIP-26) and let the different actors advertise which off-chain storage they pay attention to. 

I've updated spec to move torwards this direction!

## Certificate JSON Schema
```json
{
  "$schema": "http://json-schema.org/draft-04/schema#",
  "type": "object",
  "properties": {
    "subject": {
      "type": "string"
    },
    "type": {
      "type": "string",
      "enum": [
        "REGISTER",
        "UPDATE"
      ]
    },
    "rootHash": {
      "type": "string"
    },
    "cip26": {
      "type": "array",
      "items": [
        {
          "type": "string"
        }
      ]
    },
    "did": {
      "type": "string"
    },
    "signature": {
      "type": "object",
      "properties": {
        "r": {
          "type": "string"
        },
        "s": {
          "type": "string"
        },
        "algo": {
          "type": "string"
        },
        "pub": {
          "type": "string"
        }
      },
      "required": [
        "r",
        "s",
        "algo",
        "pub"
      ]
    }
  },
  "required": [
    "subject",
    "rootHash",
    "cip26",
    "signature"
  ]
}
```

## Metadata Label

When submitting the transaction metadata picks one of the following values for `transaction_metadatum_label`:

- `666`: DApp Registration

> - `667`: DApp Certification

Reviewer 2: ❔ Why `666` and shall we chose a less conotated number?

Author: Fair :+1:  ; Picked a less conotated number that was available according to the https://github.com/cardano-foundation/CIPs/blob/master/CIP-0010/registry.json

                    "contractAddress": {
                      "type": "string"
                    }
                  },
                  "required": [
                    "version",
                    "plutusVersion",
                    "fullScriptHash",
                    "scriptHash",
                    "contractAddress"
                  ]
                }
              ]
            }
          },
          "required": [
            "id",
            "name",
            "purpose",
            "type",
            "versions"
          ]
        }
      ]
    }
  },
  "required": [
    "subject",
    "projectName",
    "link",
    "social",
    "category",
    "subCategory",
    "description",
    "releases",
    "scripts"
  ]
}
```

### Example

```json
{
  "subject": "9SYAJPNN",
  "projectName": "My Project",
  "link": "https://myProject.app",
  "social": {
    "twitter": "twiterHandle",
    "github": "githubHandle",
    "website": "https://website"
  },
  "category": "GAMING",
  "subCategory": "RPG",
  "description": {
    "short": "A story rich game where choices matter"
  },
  "releases": [
    {
      "releaseNumber": 1,
      "releaseName": "V1",

>       "auditId": "z5L90f",

Reviewer: I thought we agreed to remove audits?

Author: Ah yes; missed this one. Thanks @matiwinnetou

---
CIP: 72
Title: Cardano DApp Registration & Discovery
Status: Proposed
Category: Metadata	
Authors: 
  - Bruno Martins <bruno.martins@iohk.io>
  - Mateusz Czeladka <mateusz.czeladka@cardanofoundation.org>
Implementors: []
Discussions:
  - https://github.com/cardano-foundation/CIPs/pull/355
Created: 2022-10-18
License: CC-BY-4.0
---

# CIP-0072: Cardano DApp Registration & Discovery

## Abstract
DApp developers do not have a standardised method to record immutable, persistent claims about their dApp(s) that their users can verify. A dApp developer needs to "register" their dApp by providing a set of claims about their dApp(s) that can be verified by the user. This CIP describes a standardised method for dApp developers to register their dApp(s) and for users to verify the claims made by dApp developers.

This proposal aims to standardise the process of dApp registration and verification, and to provide a set of claims that dApp developers can use to register their dApp(s).

> ## Motivation

Reviewer 3: Thanks!

          "version": 1
        }
      ]
    }
  ],
  "scripts": [
    {
      "id": "PmNd6w",
      "name": "Marketplace",
      "purpose": "SPEND",
      "type": "PLUTUS",
      "versions": [
        {
          "version": 1,
          "plutusVersion": 1,
          "fullScriptHash": "711dcb4e80d7cd0ed384da5375661cb1e074e7feebd73eea236cd68192",
          "scriptHash": "1dcb4e80d7cd0ed384da5375661cb1e074e7feebd73eea236cd68192",
          "contractAddress": "addr1wywukn5q6lxsa5uymffh2esuk8s8fel7a0tna63rdntgrysv0f3ms"
        }
      ]
    }
  ]
}
```

### **Stores Custom fields**
Each store might have their own requirements for the metadata. For example, some stores might require a field for logo, or screenshots links. The store's should adviertise what fields they require in their documentation so that developers are aware and they can include them in the metadata. 

### **Offchain Metadata Storage**

There are multiple options to store metadata offchain. The most common options are:
- [CIP-26](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026) compliant servers
- [IPFS](https://ipfs.tech/)
- [Bitbucket](https://bitbucket.org/)
- Any REST JSON API

> 

Reviewer 3: ```suggestion

## Rationale: how does this CIP achieve its goals?
<!-- The rationale fleshes out the specification by describing what motivated the design and what led to particular design decisions. It should describe alternate designs considered and related work. The rationale should provide evidence of consensus within the community and discuss significant objections or concerns raised during the discussion.
It must also explain how the proposal affects the backward compatibility of existing solutions when applicable. If the proposal responds to a CPS, the 'Rationale' section should explain how it addresses the CPS, and answer any questions that the CPS poses for potential solutions.-->

## Path to Active

### Acceptance Criteria
<!-- Describes what are the acceptance criteria whereby a proposal becomes 'Active' -->

### Implementation Plan
<!-- A plan to meet those criteria. Or `N/A` if not applicable. -->

## Copyright
<!-- The CIP must be explicitly licensed under acceptable copyright terms. -->
```

Please add these sections.

Reviewer: I am working on this, thanks

Reviewer 3: Thanks for adding!

                    "auditLink",
                    "auditType",
                    "signature",
                    "publicKey"
                  ]
                }
              ]
            }
          },
          "required": [
            "id",
            "name",
            "purpose",
            "type",
            "versions",
            "audits"
          ]
        }
      ]
    }
  },
  "required": [
    "subject",
    "projectName",
    "link",
    "twitter",
    "category",
    "subCategory",
    "description",
    "releases",
    "scripts"
  ]
}
```

## Example

```json
{
    "subject": "9SYAJPNN",
    "projectName": "My Project",
    "link": "https://myProject.app",
    "twitter": "https://twitter.com/MyProject",
    "category": "GAMING",
    "subCategory": "RPG",
    "description": {
                   "short": "A story rich game where choices matter"
    },	

    "releases": [
        {
            "releaseNumber": 1,
            "releaseName": "V1",
            "auditId": "z5L90f",
            "scripts": [
                {
                    "id": "PmNd6w",
                    "version": 1
                }
            ]
        }
    ],

>     "scripts": [

Author: @matiwinnetou As we discuss last i'm including a similar structure of the suggested metadata structure for review and more collaboration.

### **Stores / Auditors**
Store and auditors should be able to follow the chain and find when a new dApp registration is **anchored** on-chain. They should then perform *integrity* and *trust* validations on the dApp's certificate and metadata. 

#### **Off-chain Location Adversitment**
Each store and auditor should make public the location of their off-chain sources where they will look for the dApp's metadata based on certificates found on-chain. These can be advertised through their own API or publically available on a website.

Sample off-chain sources could be for example :
- [CIP-26](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0026) compliant servers
- IPFS

##### **Targetted Releases**
Each developer and publisher can choose to where to write metadata based on the information available from known stores & auditors. This gives **developers** and **publishers** the ability to perform targeted releases. (i.e to which stores and auditors)

#### **Suggested Validations**

- **`integrity`**: The dapp's metadata off-chain should match the metadata anchored on-chain.

> - **`trust`**: The dApp's certificate should be signed by a trusted entity. It's up to the store/auditor to decide which entities are trusted and they should maintain their own list of trusted entities. Although this entities might be well known, it's not the responsibility of this CIP to maintain this list.

Reviewer 4: ```suggestion
- **`trust`**: The dApp's certificate should be signed by a trusted entity. It's up to the store/auditor to decide which entities are trusted and they should maintain and publish their own list of trusted entities. Although this entities might be well known, it's not the responsibility of this CIP to maintain this list.
```

> - **`trust`**: The dApp's certificate should be signed by a trusted entity. It's up to the store/auditor to decide which entities are trusted and they should maintain and publish their own list of trusted entities. Although this entities might be well known, it's not the responsibility of this CIP to maintain this list.

Reviewer 3: Could this be linked to the auditors list from [CIP-0052](https://github.com/cardano-foundation/CIPs/tree/master/CIP-0052)?

> ### **On-chain dApp Registration certificate**

Reviewer 3: Is there any use case for attaching an expiration to metadata registrations? such as "after X epochs this metadata is out of date and should be updated" - Although this may introduce problems with updates being missed by chain followers who are waiting for expirations before checking chain.

```json
{
	"subject": "d684512ccb313191dd08563fd8d737312f7f104a70d9c72018f6b0621ea738c5b8213c8365b980f2d8c48d5fbb2ec3ce642725a20351dbff9861ce9695ac5db8",
	"rootHash": "8c4e9eec512f5f277ab811ba75c991d51600c80003e892e601c6b6c19aaf8a33",
  "schema_version": "0.0.1",
  "type": { 

>     "action": "REGISTER",

Author: Making it explicit, gives more information to stores when indexing updates. For example, "have i seen a previous version of this dapp, did i trusted the previous update?". Making this ambiguous, will make it more difficult for an operator to decide it's there's history of this dapp / developer

>     "releaseNumber": "1.0.0",

Author: I think the stores / auditors and all other actors should include it as part of their own validation. Any other ideas?

>     "releaseNumber": "1.0.0",

Reviewer: @Ryun1 - at first I thought that this is a good comment and we should put that versions should be always increasing for new updates but then I realised that somebody may have missed an update and wants to retrospectively fill it. 

I think it is best not to stipulate such things albeit it looks like a very good idea at first.

>     "releaseNumber": "1.0.0",

Reviewer 3: @matiwinnetou - Makes sense to me :)

    "releaseName": "My First Release",
  },
	"signature": {
		"r": "5114674f1ce8a2615f2b15138944e5c58511804d72a96260ce8c587e7220daa90b9e65b450ff49563744d7633b43a78b8dc6ec3e3397b50080",
		"s": "a15f06ce8005ad817a1681a4e96ee6b4831679ef448d7c283b188ed64d399d6bac420fadf33964b2f2e0f2d1abd401e8eb09ab29e3ff280600",
		"algo": "Ed25519−EdDSA",
		"pub": "b7a3c12dc0c8c748ab07525b701122b88bd78f600c76342d27f25e5f92444cde"
	}
}
```

### Properties

> *`subject`*: Identifier of the claim subject (i.e dapp). A UTF-8 encoded string. This uniquess of this property cannot be guaranteed by the protocol and multiple claims for the same subject may exist therefore it is required to exist some mechanism to assert trust in the *veracity* of this property.

Reviewer 4: Note that there might be multiple valid claims, e.g. for different audits of the same (version of a) DApp, presumably by different auditors; and indeed there might be multiple claims by _the same claimant_, by way of updates.

*`type`*: The type of the claim. This is a JSON object that contains the following properties: 

> - *`action`*: The action that the certificate is asserting. It can take the following values: 

Reviewer 4: Will we also add `CERTIFICATION` here? (Though there's a horrible ambiguity in usage here; not your fault.)

  - *`REGISTER`*: The certificate is asserting that the dApp is being registered for the first time. 
  - *`UPDATE`*: The certificate is asserting that the dApp is being updated.

*`rootHash`*: The hash of the metadata entire tree object. This hash is used by clients to verify the integrity of the metadata tree object. When reading a metadata tree object, the client should calculate the hash of the object and compare it with the `rootHash` property. If the two hashes don't match, the client should discard the object. The metadata tree object is a JSON object that contains the dApp's metadata. The metadata tree object is described in the next section.

> This hash is calculated by taking the entire metadata tree object, ordering the keys in the object alphanumerically, and then hashing the resulting JSON string using the blake2b-256 hashing algorithm. The hash is encoded as a hex string.

Reviewer 4: Is this specific enough to be canonical? Discussion on Stack overflow suggests that need to handle e.g. whitespace as well? https://stackoverflow.com/questions/4670494/how-to-cryptographically-hash-a-json-object

> *`signature`*: The signature of the certificate. The signature is done over the blake2b hash of the certificate. The client should use the public key to verify the signature of the certificate. 

Reviewer 4: Throughout uses both `blake2b` and `blake2b-256`; should all uses prescribe the size?

        "properties": {
          "action": {
          	"type": "string",
            "description": "Describes the action this certificate is claiming. I.e 'REGISTER', for a new dapp; or 'UPDATE' for a new release"
          },
          "releaseNumber": {
          	"type": "string",
            "description": "offical version of the release"
          },
          "releaseName": {
          	"type": "string",
            "description": "Dapp release name"
          }
		}
      },  
      "rootHash": {
        "type": "string",
        "description": "blake2b hash of the metadata describing the dApp"
      },
      "signature": {
        "description": "Age in years which must be equal to or greater than zero.",
        "type": "object",
        "properties": {
			"r": {
				"type": "string",
				"description": "hex representation of the R component of the signature"
			},
            "s": {
				"type": "string",
				"description": "hex representation of the S component of the signature"
			},
              
		},
		"required": ["r", "s", "algo", "pub"]
      }
    },

>     "required": ["subject", "rootHash","type", "signature"]

Reviewer 3: Add `schema_version` to the required? also I suggest changing `schema_version` to match how versioning was done in [CIP-25](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0025/README.md?plain=1#L67) - that it is just an integer.

Author: Interesting. Could you suggest a change to the CIP? 

CC: @matiwinnetou

Reviewer 3: This is outdated idea from me, not worth pursuing.

            "auditId": "z5L90f",
            "scripts": [
                {
                    "id": "PmNd6w",
                    "version": 1
                }
            ]
        }
    ],
    "scripts": [
        {
            "id": "PmNd6w",
	    "name": "Marketplace",
            "purpose": "SPEND",
	    "type": "PLUTUS",		
            "versions": [
                {
                    "version": 1,
		    "plutusVersion": 1, 
	            "fullScriptHash": "711dcb4e80d7cd0ed384da5375661cb1e074e7feebd73eea236cd68192",                     
                    "scriptHash": "1dcb4e80d7cd0ed384da5375661cb1e074e7feebd73eea236cd68192",
                    "contractAddress": "addr1wywukn5q6lxsa5uymffh2esuk8s8fel7a0tna63rdntgrysv0f3ms"
                }    
            ],
          "audits": [
              {
                "auditId": "z5L90f",
                "auditor": "Canonical LLC.",
                "auditLink": "https://github.com/somlinkToAessment",
                "auditType": "MANUAL",
                "signature": "0x1234567890abcdef",
                "publicKey": "0x1234567890abcdef"
              }
          ]
        }
    ]
}
```

### **Scripts Audits**

> As part of the metadata we will be including a list of audits for each script. This will be a list of audits that have been performed on the script. The audits will be performed by a third party auditor and will be signed by the auditor. 

Reviewer 4: ```suggestion
As part of the metadata we will be including a list of audits for each DApp. This will be a list of audits that have been performed on the deployed scripts comprising the DApp. The audits will be performed by a third party auditor and will be signed by the auditor. 
```

> As part of the metadata we will be including a list of audits for each script. This will be a list of audits that have been performed on the script. The audits will be performed by a third party auditor and will be signed by the auditor. 

Reviewer 4: Note that in practice this information might be provided by a _separate_ DApp registration, that simply links the audit to the deployed set of scripts comprising a DApp version.

Author: I guess here that's definitely an option. That the auditors / certification issuers emit their own cert (maybe some structure) and they are responsible for going through every  single script and metadata registered. 

The other option could be that they only look at scripts and if the developer can include the audit of every script they use; we could consider that as "certified"?

Some of the properties included in the audit are:

- **auditType** property will be included that can take one of the following values: MANUAL, AUTOMATED, or MIXED.
- **auditId** property will be included that will be a unique identifier for the audit.
- **auditLink** property will be included that will be a link to the audit report.
- **signature** property will be included that will be a signature of the audit report.
- **publicKey** property will be included that will be the public key of the auditor.

> 

Reviewer 4: ```
# Certification

This section describes how certification metadata can be included in CIP-72-compliant metadata for DApp registration. It first describes the requirements for DApp certification, and then sets out the options for its inclusion in CIP-72.

Note: In this section ‘certification’ is used to mean the process of providing various kinds of assurance of DApps through the provision of verifiable, trustworthy metadata linking to relevant evidence of assurance, such as audit reports, test run data etc.

It is expected that evidence of various kinds of assurance of DApps is recorded in an immutable and verifiable way on the Cardano blockchain; these forms of assurance are: testing, including property-based testing (level 1), audit (level 2) and formal verification (level 3). Certification is to be provided by certification providers including: testing services (level 1), auditors (level 2), and verification services (level 3).

## Requirements

Certification metadata should be related to the set of scripts that constitute the deployed version of the DApp that has been certified. 

Certification metadata should contain the following information.:

* the set of deployed scripts to which the metadata pertains
* the certification provider
* the hash of the offchain meta-data by the certification provider
* links to relevant evidence off chain, including, as appropriate,
  * source code repository
  * compiler and version used
  * parameter values for parametric validators 
  * audit report	
  * audit summary (max 100 words)
  * test run data
  * formal proof objects. 
  * certified compilation proof object, etc.

The certification metadata should be linked to DApp registration, as outlined in this CIP.  This requires the registration of any certified DApp prior to, or simultaneously with, the submission of the certification metadata. While audit and testing may take place before a DApp is registered or released, certification is only significant once the DApp is released, and so it can take place once the DApp is registered.

Certification metadata should be signed by the certification provider. It is assumed that certification providers make available their public keys for signature verification.

The metadata should be discoverable by all certification stakeholders, including end-users, DApp developers, and ecosystem components, such as light wallets and DApp stores. Information should be indexable by certification provider, DApp developer, DApp and DApp version.

It should be possible for there to be multiple versions of metadata published by the same certification provider for the same version of a DApp, particularly when a (minor) update of the evidence is necessary. In such a case it should be possible to identify the most recent version of the report for that DApp version. It will also be the case that the same DApp may have certification metadata provided by multiple different certification providers.

It should be possible for wallets to identify to users the certification status of a DApp when they are signing a transaction that is being submitted to a deployed DApp.

## Mechanisms

Evidence of certification is to be posted as explained here. Line references here relate to [CIP-0072? | DApp Registration & Discovery](https://github.com/cardano-foundation/CIPs/pull/355/files#top). 

Evidence is to be published as a separate on-chain item, but as a new `type` of CIP-72 registration.
* Use the same mechanism as for certification (lines 88–142), with a new `type`, `CERTIFICATION` (lines 77–80).
* Assumes that the `subject` will be common with the registration and update of the DApp.
* Assume that it has the same value for `transaction_metadatum_label`, 1667 (lines 143–147).
* This option requires a mechanism to link the certification data to the DApp: this is done via the subject field of the metadata (“the DApp”) and the set of deployed scripts (“the DApp version”).

Advantages
* This item can be managed by the certificate provider independently of other parties. 
* This item can be concerned with certification data only.
* This item is identified by the same label as registration data.
* The metadata will only require update if there is a change in the certification information for the given set of deployed scripts.

Disadvantages
* Requires DApp stores to implement the linking mechanism described above.

Question 
* Will using this typing mechanism require the type `UPDATE` to be split into `REG-UPDATE` and `CERT-UPDATE`?

## General discussion 
1. The model of update outlined in CIP-72, using an explicit `type` for updates, seems to be a particular version of the more general mechanism described in [CIP-26](https://cips.cardano.org/cips/cip26/). It would be clearer to adopt the CIP-26 model for managing metadata, rather than being mechanism agnostic but having to reinvent some aspects of it,
1. From the certification perspective, it makes no sense to audit single contracts. All audit and certification activity is related to DApps, that is to sets of deployed scripts. This needs to be reflected in CIP-72, which links audits to scripts, individually. Of course, it is possible to link from a single script to an audit via the set(s) of scripts to which it belongs (giving the context for the audit).
1. Certificates in CIP-72 are classified as `MANUAL`, `AUTOMATED`, `MIXED`. Preferably this should be levels 1, 2 and 3, fitting with the widely discussed approach to assurance.
1. There needs to be a canonical mechanism for identifying a set of scripts. Two options:
  * Order the script hashes, concatenate them, and then hash the result.
  * Order the scripts according to script names, hash them, concatenate and hash.

```

Reviewer 4: Suggest certification of assurance evidence as outlined above.

> ### **Stores Custom fields**

Author: I think that each store should advertise on their own these. Adding them here would make this CIP some sort of database for available stores and somewhat permission ed based on what get's merged or not.

using Datums requires a smart contract and we want to keep this solution as accessible as possible. It is a GUI concern since if there is a 
some app that can attest validity and conformance to JSON schema - dApp Registration / Update MUST never be done that does not conform to the schema.

### Scripts / Releases Fields Are Not Required
We made a decision to change the schema so that scripts and releases are no longer required. This could help to get initial registration from dApp developers faster and
some stores simply do not require dApps to add their scripts in order to be listed.

### Schema Version or CIP Version
We want to take schema version further and actually attest the whole CIP version so we are introducing a top level field: version. Any changes to the document will require bumping the version. CIP version needs to be also
propagated in the on-chain JSON itself.

### Tags
We briefly discussed tags and we will likely introduce tags in the near future. An array of tags to help stores / dApp developers categories where their dApp should show. This will complement `categories` field.

> ### DE-REGISTER

Reviewer 3: This addition raises a question for me:

- What should happen if a publisher registers then de-registers then registers again? is this a valid flow?

#### **Suggested Validations**
- **`integrity`**: The dApp's off-chain metadata should match the metadata **anchored** on-chain.
- **`trust`**: The dApp's certificate should be signed by a trusted entity. It's up to the store/auditor to decide which entities are trusted and they should maintain and publish their own list of trusted entities. Although this entities might be well known, it's not the responsibility of this CIP to maintain this list. These entities could be directly associated with developer/publisher or not.

### **On-chain dApp Registration Certificate**

The on-chain dApp registration certificate MUST follow canonical JSON and be serialised according to RFC 8785 (https://www.rfc-editor.org/rfc/rfc8785). This stipulation is to avoid any ambigiutines in the signature calculation.

```json
{
  "subject": "b37aabd81024c043f53a069c91e51a5b52e4ea399ae17ee1fe3cb9c44db707eb",
  "rootHash": "7abcda7de6d883e7570118c1ccc8ee2e911f2e628a41ab0685ffee15f39bba96",
  "metadata": [

>     "https://cip26.foundation.app/my_dapp.json"

Reviewer 5: I feel like **cip26.** prefix can be removed from the link @matiwinnetou

Reviewer: I already removed it, thanks

---
CIP: 72
Title: Cardano dApp Registration & Discovery
Status: Proposed
Category: Metadata
Authors: 
  - Bruno Martins <bruno.martins@iohk.io>
  - Mateusz Czeladka <mateusz.czeladka@cardanofoundation.org>

> Implementors: []

Reviewer 6: Given that there's an implementation plan, this shouldn't be empty.

Reviewer: done

                    "properties": {
                      "id": {
                        "type": "string"
                      },
                      "version": {
                        "type": "integer"
                      }
                    },
                    "required": [
                      "id",
                      "version"
                    ]
                  }
                ]
              }
            },
            "required": [
              "releaseNumber"
            ]
          }
        ]
      },
      "scripts": {
        "type": "array",
        "items": [
          {
            "type": "object",
            "properties":{
              "id": {
                "type":"string",
                "description": "Unique Script ID (across all scripts from this dApp)."
              },
              "name":{
                "type":"string",
                "description": "An optional script name usually related to it's function."
              },
              "purposes":{
                "type":"array",
                "items": {
                    "type": "string",
                      "enum":["SPEND", "MINT"]
                },

>                 "description": "Purpouses of the script, SPEND or MINT (notice it can be both for some modern Cardano languages)."

Reviewer 3: ```suggestion
                "description": "Purposes of the script, SPEND or MINT (notice it can be both for some modern Cardano languages)."
```
This misspelling happens a few times :)

### Who Is The Owner?
Smart contracts are ownerless, it has been debated that there could be multiple claims to the same dApps from different parties.

The standard doesn't prevent anyone from making a claim, so it's up to the different operator to their diligence work and make their own choices of whom they trust. The signature should give the most confidence as anyone can collect known public keys from known development companies. Future CIP revisions can include DID's and Verifiable Credentials to tackle this problem in a more elegant way.

### DIDs
Since DIDs / Verifiable Credetials are not yet widely used in Cardano ecosystem, usage of them in this initial CIP version has been de-scoped.

### Categories
`Categories` is a predefined enum with values defined in the CIP / schema, it is *NOT* a free text field, rationale for this is that dApp developers will have no idea what ontology / classification to use, which will likely result in many duplicates of the same thing.

> ### Purpouse Field As an Array or as a Single Item?

Reviewer 6: ```suggestion
### Purpose Field As an Array or as a Single Item?
```

It may have been visible that we have a `purpose` field, which can be: "SPEND" or "MINT", those fields directly map to what is allowed by a Cardano Smart Contract. As of the time of writing CIP - PlutusTx does not allow a script to be both of type: "SPEND" and "MINT", however, there are new

> languages on Cardano being worked on where they already allow one validator to be both spending UTxOs and minting tokens - all with the same script hash. To be open for the future it has been agreed to turn `purpouse` field into `purpouses` and make it a JSON array.

Reviewer 6: ```suggestion
languages on Cardano being worked on where they already allow one validator to be both spending UTxOs and minting tokens - all with the same script hash. To be open for the future it has been agreed to turn `purpose` field into `purposes` and make it a JSON array.
```

### Version Deprecation
We discussed scenario what to do in case a dApp team wants to deprecate a particular version. Upon a few iteration we settled on doing this in off-chain section.

### Version Security Vulnerability Flagging
It is not uncommon to see a dApp release a version and then release a fix in the new version and flag the previous version
as having security vulnerability. We are intoducing an optional field in the offchain json on the release level: `securityVulnerability": true.

### Comment Field (on-chain JSON)
We are introducing a field in the on-chain JSON only, which allows dApp development teams to provide a free text field
comment about changes they are making in a given (re-)registration request.

## Path to Active
We will evaluate for a few months if we have not missed any details, collect feedback from dApp developers, stores. We reserve right to make small changes in this time, while the proposal
is in `PROPOSED` status. After certain period (no longer than 6 months), from time of merging PR in `PROPOSED` status to the main / master branch, we will

> update the proposal to be in `ACTIVE` state.

Reviewer 6: This can only happen if the proposal meets its acceptance criteria though.

Reviewer: makes sense, changed

> ### Acceptance Criteria

Reviewer 6: As a concrete acceptance criteria, I'd suggest to have at least 2-3 non-trivial dApps **from different entities** registered through this method. That would serve as a good measure of its acceptance.

Reviewer: done

> - IOG and CF approval

Reviewer 3: How is this measured?

Reviewer 6: More importantly, this CANNOT be an acceptance criteria. IOG and CF are not authorities when it comes to approving CIPs. Acceptance criteria must reflect actually measurable data points and show "proof of activity / usage".

> - Community representative approval (Santiago / TxPipe and Marcel / Eternl wallet)

Reviewer 6: If these people are meant to be implementors of the standard, then we should list them as implementors. And, it goes without saying that **they** should confirm having such role.

Reviewer: They didn't, I removed them.

- PR to https://github.com/cardano-foundation/CIPs/blob/master/CIP-0010/registry.json

### Implementation Plan
- Reference implementation: https://github.com/Cardano-Fans/crfa-dapp-registration-and-certification-service (incubator)

## Copyright

> [CC-BY-4.0]: https://creativecommons.org/licenses/by/4.0/legalcode

Reviewer 3: ```suggestion
[CC-BY-4.0](https://creativecommons.org/licenses/by/4.0/legalcode)
```

> [Apache-2.0]: http://www.apache.org/licenses/LICENSE-2.0

Reviewer 3: ```suggestion
```
Removed to match copyright in the header

*`metadata`*: An array of links to the dApp's metadata. The metadata is a JSON compatible RFC 8785 object that contains the dApp's metadata.

*`signature`*: The signature of the certificate. The publishers generate the signature is by first turning on-chain JSON into a canonical form (RFC 8765), hashing it with blake2b-256 and generating a signature of the hash. Stores / clients can verify the signature by repeating the process, they can use the public key to verify the signature of the certificate. Fields used for canonical JSON: ["subject", "rootHash", "metadata","type"]. Please note that a signature should be generated of blake2b-256 hash as a byte array, not as a hex represented string(!).

### On chain CDDL for registration / de-registration
```
string = bstr .size (1..64) ; tstr / string from 1 up to 64 chars only

sig_256 = bstr .size (64..64) ; 256 bytes signature (256 / 64 = 4 bytes)

transaction_metadata = {
  1667: on-chain_metadata
}

on-chain_metadata = {
  subject: string,
  rootHash: sig_256,
  metadata: [+ string] / [+ string / [+ string]],
  type: registration / de-registration,
  signature: signature,
}

registration = {

>   action: "REGISTRATION",

Reviewer 7: ```suggestion
  action: "REGISTER",
```

         "description": "An array of valid URLs pointing to off-chain CIP-72 compatible metadata document. If an individual URL is longer than 64 characters, it must be expressed as an array of strings (where each string may contain at most 64 characters).",
         "items": {
           "anyOf": [{
             "type": "string",
               "minLength": 1,
               "maxLength": 64
             }, {
             "type": "array",
             "items": {
               "type": "string",
               "minLength": 1,
               "maxLength": 64
             }
           }],
           "examples": ["https://raw.githubusercontent.com/org/repo/offchain.json", ["https://raw.githubusercontent.com/long-org-name/", "long-repo-name/offchain.json"], "ipfs://QmbQDvKJeo2NgGcGdnUiUFibTzuKNK5Uij7jzmK8ZccmWp", ["ipfs://QmbQDvKJeo2NgGcGdnUiaAdADA", "UFibTzuKNKc0jA390alDAD5Uij7jzmK8ZccmWp"]]
         }
       },
       "type":{
          "type":"object",
          "description":"Describes the releases, if they are new or an updates.",
          "properties":{
             "action":{
               "type":"string",
               "enum":["REGISTER", "DE_REGISTER"],

>                "pattern": "^(REGISTER|DE_REGISTER)$",

Reviewer 7: If this is defined as enum, do we additionally need the regex?

Reviewer: probably not, good point

            "minLength": 40,
            "maxLength": 168
          },
          "long": {
            "type": "string",
            "description": "An optional long dApp description (no less than 40 and no longer than 1008 characters).",
            "minLength": 40,
            "maxLength": 1008
          }
        },
        "required": [
          "short"
        ]
      },
      "releases": {
        "type": "array",
        "items": [
          {
            "type": "object",
            "properties": {
              "releaseNumber": {
                "type": "string",
                "pattern": "^(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)\\.(0|[1-9][0-9]*)(-((0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*)(\\.(0|[1-9][0-9]*|[0-9]*[a-zA-Z-][0-9a-zA-Z-]*))*))?(\\+([0-9a-zA-Z-]+(\\.[0-9a-zA-Z-]+)*))?$",
                "description": "Semver compatible release number (major.minor.patch) or (major.minor.patch-some_text) e.g. 1.2.3 or 1.1.1-alpha"
              },
              "releaseName": {
                "type": "string",
                "description": "An optional human readable release name, e.g. V1"
              },

>               "security_vulnerability": {

Reviewer 7: Small suggestion - we are naming other properties with camelCase:
```suggestion
              "securityVulnerability": {
```
